x/tx/aminojson/json_marshal.go

```diff
@@ -0,0 +1,357 @@
+	}
+
+	if encoder := aj.getMessageEncoder(msg); encoder != nil {
+		err := encoder(msg, writer)
+		return err
+	}
+
+	_, err := writer.Write([]byte("{"))
+	if err != nil {
+		return err
+	}
+
+	fields := msg.Descriptor().Fields()
+	first := true
+	emptyOneOfWritten := map[string]bool{}
+	for i := 0; i < fields.Len(); i++ {
+		f := fields.Get(i)
+		v := msg.Get(f)
+		name := getFieldName(f)
+		oneof := f.ContainingOneof()
+		isOneOf := oneof != nil
+		oneofFieldName, oneofTypeName, err := getOneOfNames(f)
+		if err != nil && isOneOf {
+			return err
+		}
+		writeNil := false
+
+		if !msg.Has(f) {
+			// msg.WhichOneof(oneof) == nil: no field of the oneof has been set
+			// !emptyOneOfWritten: we haven't written a null for this oneof yet (only write one null per empty oneof)
+			if isOneOf && msg.WhichOneof(oneof) == nil && !emptyOneOfWritten[oneofFieldName] {
+				name = oneofFieldName
+				writeNil = true
+				emptyOneOfWritten[oneofFieldName] = true
+			} else if omitEmpty(f) {
+				continue
+			} else if f.Kind() == protoreflect.MessageKind &&
+				f.Cardinality() != protoreflect.Repeated &&
+				!v.Message().IsValid() {
+				return errors.Errorf("not supported: dont_omit_empty=true on invalid (nil?) message field: %s", name)
+			}
+		}
+
+		if !first {
+			_, err = writer.Write([]byte(","))
+			if err != nil {
+				return err
+			}
+		}
+
+		if isOneOf && !writeNil {
+			_, err = writer.Write([]byte(fmt.Sprintf(`"%s":{"type":"%s","value":{`,
+				oneofFieldName, oneofTypeName)))
+			if err != nil {
+				return err
+			}
+		}
+
+		err = jsonMarshal(writer, name)
+		if err != nil {
+			return err
+		}
+
+		_, err = writer.Write([]byte(":"))
+		if err != nil {
+			return err
+		}
+
+		// encode value
+		if encoder := aj.getFieldEncoding(f); encoder != nil {
+			err = encoder(aj, v, writer)
+			if err != nil {
+				return err
+			}
+		} else if writeNil {
+			_, err = writer.Write([]byte("null"))
+			if err != nil {
+				return err
+			}
+		} else {
+			err = aj.marshal(v, writer)
+			if err != nil {
+				return err
+			}
+		}
+
+		if isOneOf && !writeNil {
+			_, err = writer.Write([]byte("}}"))
+			if err != nil {
+				return err
+			}
+		}
+
+		first = false
+	}
+
+	_, err = writer.Write([]byte("}"))
+	return err
+}
+
+func jsonMarshal(w io.Writer, v interface{}) error {
+	blob, err := json.Marshal(v)
+	if err != nil {
+		return err
+	}
+	_, err = w.Write(blob)
+	return err
+}
+
+func (aj AminoJSON) marshalList(list protoreflect.List, writer io.Writer) error {
+	n := list.Len()
+
+	_, err := writer.Write([]byte("["))
+	if err != nil {
+		return err
+	}
+
+	first := true
+	for i := 0; i < n; i++ {
+		if !first {
+			_, err := writer.Write([]byte(","))
+			if err != nil {
+				return err
+			}
+		}
+		first = false
+
+		err = aj.marshal(list.Get(i), writer)
+		if err != nil {
+			return err
+		}
+	}
+
+	_, err = writer.Write([]byte("]"))
+	return err
+}
+
+func getMessageName(msg protoreflect.Message) (string, bool) {
```

Comment:
Would add a godoc here, or else it could be thought that this funciton returns `msg.Name()`